test/common/thread_local/thread_local_impl_test.cc

```diff
@@ -82,6 +82,40 @@ TEST_F(ThreadLocalInstanceImplTest, All) {
   tls_.shutdownThread();
 }
 
+// Validate ThreadLocal::runOnAllThreads behavior with all_thread_complete call back.
+TEST_F(ThreadLocalInstanceImplTest, RunOnAllThreads) {
+
+  SlotPtr tlsptr = tls_.allocateSlot();
+
+  EXPECT_CALL(thread_dispatcher_, post(_));
+  EXPECT_CALL(main_dispatcher_, post(_));
+
+  // Ensure that the thread local call back and all_thread_complete call back are called.
+  struct thread_local_verification_info {
+    std::atomic<uint64_t> thread_local_calls_{0};
+    std::condition_variable condvar_;
+    std::mutex condvar_mutex_;
+    bool all_threads_complete_;
+  };
+
+  thread_local_verification_info thread_local_data;
+
+  tlsptr->runOnAllThreads(
+      [&thread_local_data]() -> void { ++thread_local_data.thread_local_calls_; },
+      [&thread_local_data]() -> void {
+        EXPECT_EQ(thread_local_data.thread_local_calls_, 1);
+        std::unique_lock<std::mutex> lock(thread_local_data.condvar_mutex_);
+        thread_local_data.all_threads_complete_ = true;
+        thread_local_data.condvar_.notify_one();
+      });
+  std::unique_lock<std::mutex> lock(thread_local_data.condvar_mutex_);
```

Comment:
put the mutex lock and the wait into a scope, so the lock is released before you call shutdownGlobalThreading.  This sounds familiar, right?  I am never sure if all my github comments are getting through.

In general hold locks for the minimum amount of time.

Reply:
i did not get this specific comment. Will change now.

Reply:
The comment was "just hold this over the cv_.wait call and not during the shutdown calls." and your response was "done. Must be sleeping when i coded this".

If you could iterate through the PR and look at all the comments exposed by "show outdated" links it would reduce the number of comments that wind up getting repeated.